source/websocket.c

```diff
@@ -0,0 +1,908 @@
+        AWS_LOGF_ERROR(
+            AWS_LS_HTTP_WEBSOCKET,
+            "id=%p: Unexpected error while calculating message size, closing websocket.",
+            (void *)websocket);
+
+        aws_raise_error(AWS_ERROR_UNKNOWN);
+        goto error;
+    }
+    io_msg_hint -= upstream_overhead;
+
+    io_msg = aws_channel_acquire_message_from_pool(
+        websocket->channel_slot->channel, AWS_IO_MESSAGE_APPLICATION_DATA, io_msg_hint);
+    if (!io_msg) {
+        AWS_LOGF_ERROR(
+            AWS_LS_HTTP_WEBSOCKET,
+            "id=%p: Failed acquire message from pool, error %d (%s).",
+            (void *)websocket,
+            aws_last_error(),
+            aws_error_name(aws_last_error()));
+        goto error;
+    }
+
+    io_msg->user_data = websocket;
+    io_msg->on_completion = s_io_message_write_completed;
+
+    /* Loop through frames, writing their data into the io_msg */
+    bool wrote_close_frame = false;
+    while (!websocket->thread_data.is_writing_stopped) {
+        if (websocket->thread_data.current_outgoing_frame) {
+            AWS_LOGF_TRACE(
+                AWS_LS_HTTP_WEBSOCKET,
+                "id=%p: Resuming write of frame=%p opcode=%" PRIu8 "(%s) payload-length=%" PRIu64 ".",
+                (void *)websocket,
+                (void *)websocket->thread_data.current_outgoing_frame,
+                websocket->thread_data.current_outgoing_frame->def.opcode,
+                aws_websocket_opcode_str(websocket->thread_data.current_outgoing_frame->def.opcode),
+                websocket->thread_data.current_outgoing_frame->def.payload_length);
+
+        } else {
+            /* We're not in the middle of encoding a frame, so pop off the next one to encode. */
+            if (aws_linked_list_empty(&websocket->thread_data.outgoing_frame_list)) {
+                AWS_LOGF_TRACE(AWS_LS_HTTP_WEBSOCKET, "id=%p: No more frames to write.", (void *)websocket);
+                break;
+            }
+
+            struct aws_linked_list_node *node = aws_linked_list_pop_front(&websocket->thread_data.outgoing_frame_list);
+            websocket->thread_data.current_outgoing_frame = AWS_CONTAINER_OF(node, struct outgoing_frame, node);
+
+            struct aws_websocket_frame frame = {
+                .fin = websocket->thread_data.current_outgoing_frame->def.fin,
+                .opcode = websocket->thread_data.current_outgoing_frame->def.opcode,
+                .payload_length = websocket->thread_data.current_outgoing_frame->def.payload_length,
+            };
+
+            /* RFC-6455 Section 5.3 Client-to-Server Masking
+             * Clients must mask payload with key derived from an unpredictable source of entropy. */
+            if (!websocket->is_server) {
+                frame.masked = true;
+                /* TODO: faster source of random (but still seeded by device_random) */
+                struct aws_byte_buf masking_key_buf = aws_byte_buf_from_empty_array(frame.masking_key, 4);
```

Comment:
I wish we had C++ here because a merseine twister is probably what we want here. Maybe we should implement one in aws-c-common

Reply:
Yup, hence the TODO